### ISSUE TITLE
I'd like to use (require) a Node module in the prelude

### ISSUE THREAD

**Issue by thorn0** [bterlson/test262-harness#45]
I'd like to load a Node module in the prelude, but things like `require`, `global`, etc. don't exist there. Would be great to have this possibility. (Currently, I'm (concatenating files)[https://github.com/thorn0/timezoned-date/blob/master/package.json] to create a working prelude).

**Comment by bterlson** [comment]
The difficulty here is having require do the right thing inside a new VM context since you can't just pass in the parent context's require function...

**Comment by thorn0** [comment]
The problem is that the parent context's `require` would look for modules at the wrong path, right?

**Comment by bterlson** [comment]
Possibly. I assume that if the module modified globals it would modify the parent realm too.

**Comment by ljharb** [comment]
ping @bterlson - this is the major barrier to being able to use test262-harness inside a polyfill/shim module as part of normal tests.

**Comment by rwaldron** [comment]
@ljharb do you have any thoughts/suggestions for how to address the issues discussed above?

**Comment by ljharb** [comment]
No, I'm not sure; but i believe you could use something like https://www.npmjs.com/package/require-for to pass a `require` into `runInNewContext` that was scoped specifically to the dirname of the prelude file?

**Comment by ljharb** [comment]
Linking to #250.

**Comment by jugglinmike** [comment]
@thorn0 @ljharb Would it be possible for you folks to use a tool like Webpack to bundle your polyfills into a string that could be injected via the `prelude` option? And in your experience, do you think this approach would generalize for any polyfill?

I'm asking this not (just) because I'm lazy but because that approach is host-independent. Polyfill authors could use it to vet their libraries across hosts--not just Node.js.

**Comment by ljharb** [comment]
It might be possible but it would be inordinately complex, and the goal is to make using the harness easy.

If the hostType is node, I would assume I could access node features properly in the prelude.

**Comment by astlaurent1** [comment]
I was attempting to run a test against the [promise.allsettled shim](https://github.com/ljharb/Promise.allSettled) and ran into this issue. `require` was defined, but it could not resolve the module

I went with @jugglinmike's proposed solution of using Webpack to bundle what I needed into a prelude file. The webpack config and bundled file is here: https://github.com/astlaurent1/Promise.allSettled/commit/631acd603fd556ab8005165c785a64faf71f0cbb. I suppose I could have done that in my fork of test262

After doing all that, in hindsight, it would seem relatively straightforward to support something like `--webpack-prelude <prelude-entry.js>`, where `<prelude-entry.js>` imports whatever modules it needs to and runs necessary bootstrapping code. In this case, the `<prelude-entry.js>` imports the shim, then calls the `.shim()` method to make `Promise.allSettled` available in the context of the test drivers

Pre-packaging the prelude like this also serves the added benefit of  not "polluting" test cases with module/import semantics. The bundling provides a nice boundary between what you're preparing and what you're testing

(P.S. See https://github.com/tc39/test262/issues/2112 for discussions on testing `Promise.allSettled`)

**Comment by ptomato** [comment]
> Would it be possible for you folks to use a tool like Webpack to bundle your polyfills into a string that could be injected via the `prelude` option? And in your experience, do you think this approach would generalize for any polyfill?

[This is what we're doing for the polyfill in proposal-temporal.](https://github.com/tc39/proposal-temporal/blob/main/polyfill/ci_test.sh) It works, although it makes test runs painfully slow (I suspect) because the prelude has to be parsed and executed for every test file.
